python/ray/dag/compiled_dag_node.py

```diff
@@ -150,54 +138,58 @@ def _wrap_exception(exc):
     return wrapped
 
 
-def _exec_task(self, task: "ExecutableTask", idx: int) -> bool:
+def _select_next_nodes(
+    actor_to_candidates: Dict["ray.actor.ActorHandle", List[DAGOperationGraphNode]],
+    graph: Dict[int, Dict[DAGNodeOperationType, DAGOperationGraphNode]],
+):
     """
-    Execute the task.
+    This function selects the next nodes for topological sort to generate execution
+    schedule. If there are multiple DAGOperationGraphNodes with zero in-degree,
+    select nodes based on the following rules:
+
+    #1  If the nodes are not NCCL write nodes, select the one with the smallest
+        `bind_index`. If there are multiple candidate nodes with the smallest
+        `bind_index` of the actors that they belong to, any one of them is
+        acceptable. For the implementation details, we maintain a priority queue
+        for each actor, where the head of the priority queue is the node with the
+        smallest `bind_index`.
+    #2  If #1 cannot be satisfied, it means that all candidate nodes are
+        NCCL write nodes. In this case, select the one that is the head of the
+        priority queue and its downstream nodes, regardless of whether the
+        downstream nodes are heads of their priority queues or not.
+
+    This function may return multiple nodes if they are NCCL nodes. In that case,
+    this function only removes the NCCL write node, which is also the head of a
+    priority queue. Other nodes will be removed in the following iterations.
```

Comment:
> Other nodes will be removed in the following iterations.

this is a little confusing. is it difficult (or expensive) to fix it?

Reply:
We can remove the downstream nodes from their priority queue when we pop them, but this is inefficient (`O(n)`).

Reply:
I can update it if `O(n)` is acceptable.

Reply:
let's keep it in that case. O(n) sounds not great (especially when there are multiple of them)

Reply:
I reconsidered this and realized that the NCCL write node's downstream nodes (i.e., NCCL read nodes) will not be in actor_to_candidates, so we can't remove them.

Reply:
Related thread: https://github.com/ray-project/ray/pull/46911#discussion_r1702396288 

Move the discussion to this thread.